### ISSUE TITLE
docker-compose setup port 4566 not reachable while old ports still work

### ISSUE THREAD

**Issue by jrm2k6** [localstack/localstack#2615]
# Type of request: This is a ...

[x] bug report
[ ] feature request

# Detailed description
Running in docker-compose setup:

localstack image: 0.11.2

```
version: '3'
services:
  localstack:
    image: localstack/localstack:0.11.2
    restart: always
    ports:
      - "4566-4599:4566-4599"
      - "8010:8010"
    environment:
      - LOCALSTACK_SERVICES=iam,s3,sns,sqs
      - DEBUG=1
      - DATA_DIR=/tmp/localstack/data
      - PORT_WEB_UI=8010
      - LAMBDA_EXECUTOR=docker
      - DOCKER_HOST=unix:///var/run/docker.sock
      - DEFAULT_REGION=us-west-2
    volumes:
      - "${TMPDIR:-/tmp/localstack}:/tmp/localstack"
```

In a different container:
`aws --endpoint-url=http://localstack:4566 sns create-topic --name MyNewTopic`

## Expected behavior

A new topic is created:
```
{
    "TopicArn": "arn:aws:sns:us-west-2:000000000000:MyNewTopic"
}
```

## Actual behavior

```
It hangs and nothing happens
```

Note: Running `aws --endpoint-url=http://localstack:4575 sns create-topic --name MyNewTopic` works just fine from the same container.

**Comment by jrm2k6** [comment]
Note: it seems to be pretty inconsistent as I was able to have it working against port 4566 once. Then, I had to update one of my container script, and did a docker-compose up again, and it is back to hanging. I also can see the message:
`Waiting for all LocalStack services to be ready` being outputted.

**Comment by whummer** [comment]
Thanks for reporting this issue @jrm2k6 . Not entirely sure how we should go about debugging this issue. You mentioned that the behavior is inconsistent - have you been able to determine whether this is (1) an issue with one of the APIs not starting up properly, or (2) a networking issue where the requests are not routed correctly from the host through to the Docker container?

>I also can see the message: Waiting for all LocalStack services to be ready being outputted.

Are you seeing this message being outputted a few times only, or is it continuously being printed in an infinite loop (I believe the default retry timeout is 7 seconds), without ever converging to a steady state?

Any details or hints on how to reproduce the issue would be appreciated. Thanks

**Comment by jrm2k6** [comment]
The `Waiting for all LocalStack services to be ready` is continuously being printed in an infinite loop.
The docker-compose setup could be the issue.
I mount a specific directory to enable persistence for S3/SQS as indicated in the README through the DATA_DIR environment variable. Another container rely on TCP 4566 to be reachable to also start a setup script in charge of creating the buckets, SNS topics and SQS queues we need. I am wondering if there is a race condition or some issues when it tries to reapply some of the API calls at the same time as this setup script, if the setup script has already attempted to create those resources.
It seems to be working fine if I remove the DATA_DIR env variables, which means we would have to always recreate the needed resources on each docker-compose up.

**Comment by NeilJ247** [comment]
I am seeing the same issue.

Config:

```
  aws-local:
    image: localstack/localstack:0.11.2
    networks:
      - my-network
    ports:
      - "4567:4566"
    environment:
      - SERVICES=sqs
      - DEFAULT_REGION=eu-west-1
      - PORT_WEB_UI=8090
      - DEBUG=1
      - DATA_DIR=/tmp/localstack/data
      - HOSTNAME_EXTERNAL=aws-local
    volumes:
      - './.localstack:/tmp/localstack'
```

In my case I use SQS only and create a worker queue and dead letter queue in a dev entrypoint script from another service.  If I am starting fresh there are no issues.  However, if I stop the containers and then restart them I see `Waiting for all LocalStack services to be ready` continually repeated after a short while which affects my worker service from processing messages.  In most cases it can process one message then repeats that message

Steps:

1. Run docker-compose up
```
aws-local_1    | 15:24:29.766 [elasticmq-akka.actor.default-dispatcher-8] INFO  o.elasticmq.actor.QueueManagerActor - Creating queue QueueData(DeadQueue-Test,MillisVisibilityTimeout(30000),PT0S,PT0S,2020-07-07T15:24:29.686Z,2020-07-07T15:24:29.686Z,None,false,false,None,None,Map())
aws-local_1    | 15:24:29.975 [elasticmq-akka.actor.default-dispatcher-9] INFO  o.elasticmq.actor.QueueManagerActor - Creating queue QueueData(Queue-Test,MillisVisibilityTimeout(30000),PT0S,PT0S,2020-07-07T15:24:29.974Z,2020-07-07T15:24:29.974Z,Some(DeadLettersQueueData(DeadQueue-Test,1)),false,false,None,None,Map())
```
2. Stop the docker-compose up
3. Run docker-compose up
```
aws-local_1    | 15:42:20.036 [elasticmq-akka.actor.default-dispatcher-10] INFO  o.elasticmq.actor.QueueManagerActor - Creating queue QueueData(DeadQueue-Test,MillisVisibilityTimeout(30000),PT0S,PT0S,2020-07-07T15:42:19.930Z,2020-07-07T15:42:19.930Z,None,false,false,None,None,Map())
aws-local_1    | 15:42:20.259 [elasticmq-akka.actor.default-dispatcher-5] INFO  o.elasticmq.actor.QueueManagerActor - Creating queue QueueData(Queue-Test,MillisVisibilityTimeout(30000),PT0S,PT0S,2020-07-07T15:42:20.258Z,2020-07-07T15:42:20.258Z,Some(DeadLettersQueueData(DeadQueue-Test,1)),false,false,None,None,Map())
aws-local_1    | Waiting for all LocalStack services to be ready
worker | 2020-07-07 16:42:22,088 INFO spawned: 'events_consumer' with pid 30
worker | 2020-07-07 16:42:23,091 INFO success: events_consumer entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
aws-local_1    | Waiting for all LocalStack services to be ready
aws-local_1    | Waiting for all LocalStack services to be ready
aws-local_1    | Waiting for all LocalStack services to be ready
```

This is the output of `localstack_infra.log`  on the first run:

```
cat localstack_infra.log 
(. .venv/bin/activate; exec bin/localstack start --host)
LocalStack version: 0.11.2
Starting local dev environment. CTRL-C to quit.
!WARNING! - Looks like you have configured $LAMBDA_REMOTE_DOCKER=1 - please make sure to configure $HOST_TMP_FOLDER to point to your host's $TMPDIR
Starting edge router (https port 4566)...
Starting mock SQS service in http ports 4566 (recommended) and 4576 (deprecated)...
15:47:51.403 [main] INFO  org.elasticmq.server.Main$ - Starting ElasticMQ server (0.15.5) ...
15:47:52.383 [elasticmq-akka.actor.default-dispatcher-5] INFO  akka.event.slf4j.Slf4jLogger - Slf4jLogger started
15:47:54.423 [elasticmq-akka.actor.default-dispatcher-6] INFO  o.e.rest.sqs.TheSQSRestServerBuilder - Started SQS rest server, bind address 0.0.0.0:56291, visible server address http://3fc25f3c706e:4576
15:47:54.426 [main] INFO  org.elasticmq.server.Main$ - === ElasticMQ server (0.15.5) started in 3956 ms ===
Ready.
```

On the second run:

```
cat /tmp/localstack_infra.log 
(. .venv/bin/activate; exec bin/localstack start --host)
LocalStack version: 0.11.2
Starting local dev environment. CTRL-C to quit.
!WARNING! - Looks like you have configured $LAMBDA_REMOTE_DOCKER=1 - please make sure to configure $HOST_TMP_FOLDER to point to your host's $TMPDIR
Starting edge router (https port 4566)...
Starting mock SQS service in http ports 4566 (recommended) and 4576 (deprecated)...
16:04:06.280 [main] INFO  org.elasticmq.server.Main$ - Starting ElasticMQ server (0.15.5) ...
16:04:07.345 [elasticmq-akka.actor.default-dispatcher-5] INFO  akka.event.slf4j.Slf4jLogger - Slf4jLogger started
16:04:09.308 [elasticmq-akka.actor.default-dispatcher-7] INFO  o.e.rest.sqs.TheSQSRestServerBuilder - Started SQS rest server, bind address 0.0.0.0:46575, visible server address http://3fc25f3c706e:4576
16:04:09.311 [main] INFO  org.elasticmq.server.Main$ - === ElasticMQ server (0.15.5) started in 3979 ms ===
16:04:10.441 [elasticmq-akka.actor.default-dispatcher-5] INFO  o.elasticmq.actor.QueueManagerActor - Creating queue QueueData(DeadQueue-Test,MillisVisibilityTimeout(30000),PT0S,PT0S,2020-07-07T16:04:10.334Z,2020-07-07T16:04:10.334Z,None,false,false,None,None,Map())
16:04:10.688 [elasticmq-akka.actor.default-dispatcher-7] INFO  o.elasticmq.actor.QueueManagerActor - Creating queue QueueData(Queue-Test,MillisVisibilityTimeout(30000),PT0S,PT0S,2020-07-07T16:04:10.686Z,2020-07-07T16:04:10.686Z,Some(DeadLettersQueueData(DeadQueue-Test,1)),false,false,None,None,Map())
```

So you can see that the `Ready` text is not present in that log.  I suspect however you guys restart the services it is not adding that bit of text into the log file.

The only way this is resolved is by either:
- removing the .localstack directory
- don't set the DATA_DIR environment variable

**Comment by jrm2k6** [comment]
@NeilJ247 thanks for investigating more. I wanted to get to it but didn't have a chance. Did you have a chance to downgrade to 0.10.* and see if the same thing is happening?

**Comment by NeilJ247** [comment]
> @NeilJ247 thanks for investigating more. I wanted to get to it but didn't have a chance. Did you have a chance to downgrade to 0.10.* and see if the same thing is happening?

Sorry no I haven't.  For the time being I am not setting the DATA_DIR.

**Comment by NeilJ247** [comment]
```
volumes:
      - './.localstack:/tmp/localstack'
```

You need to remove this volume mount and the directory in order for the problem to be resolved.

**Comment by jrm2k6** [comment]
Yep thats what I do. Will try to investigate more this weekend.

**Comment by jrm2k6** [comment]
OK, I was able to dig a bit more. Here is what I found:
The setup I was using with my original problem is very similar to this: https://github.com/jrm2k6/playground-localstack/blob/master/docker-compose.yml

So a few things:
- I was using dockerize to wait until 4566 is available on TCP. But it looks like dockerize doesnt release the TCP connection until the next command is finished, which in the case of a long running command, will never happen (https://github.com/jwilder/dockerize/blob/9ce2e80619962abe9ba90619e009d29943446ab3/main.go#L174)
- Switching to http://localstack:4566/health would return 404 when using 0.11.2. Switching to 0.11.3 fixes it.

The reason it was also hanging on Waiting for localstack services to be ready is probably because it was not able to re-run the list of API calls. So I am not sure my issue had anything to do with DATA_DIR anymore.

**Comment by NeilJ247** [comment]
Hi @jrm2k6,

Sorry to come back.  So I have switched to the latest version and observed the following:

- My docker entrypoint script uses [waitforit](https://github.com/vishnubob/wait-for-it), similar to your approach, to wait for the port to be available.  Example: `waitforit -address=http://aws-local:4576 -timeout=1200 -debug -- bin/console app:create:queue ${EVENTS_QUEUE} ${DEAD_EVENTS_QUEUE}`
- If I start localstack on its own it works fine
- If I start up another service along with localstack that hits the `http://aws-local:4576` (such as the command above) the localstack services never get into a "Ready" state
- I have switched the `waitforit` command address to http://aws-local:4576/health but this doesn't fix the issue for me

It feels like that if anything is hitting that port it whilst the services are being initialised it just screws it up.

I am going to try switching to `dockerizez and see if that resolves the issue for me in the time being.

**Comment by jrm2k6** [comment]
@NeilJ247 yes, I also experimented with waitforit but switch to dockerize as it was more straightforward and gave me more flexibility. Can you reproduce the issue if you use tcp 4576/4566 instead of http? I think dockerize doesn't close the tcp connection so it makes it inaccessible by other services trying to access 4576/4566.

**Comment by NeilJ247** [comment]
I've tried tcp and it works (that is if you mean changing dockerize command to use tcp?).

There is still an issue with `Waiting for all LocalStack services to be ready` being repeated if restart the service.  The only way to resolve that is to remove the container before starting the service again.  I can raise this as a separate issue?

**Comment by jrm2k6** [comment]
Yeah, might be better. I think using tcp with dockerize seems to make localstack unaccessible for some reason.

**Comment by ktamas77** [comment]
Hey everyone, as of today with the latest localstack (v0.12.3) the problem still 'persists':

1. Starting up localstack with persistent local volume mapping works - at the first time
2. Whenever i stop the containers & restart them, Localstack enters to an infinite loop.
3. If I delete the external volume between restarts, it works again fine.

**Comment by whummer** [comment]
Hi @ktamas77 , thanks for the update. Can you please share the tag/age of the image you're using (I'm assuming you pulled the `latest` image, not the tagged version)?
```
docker images | grep localstack
```
Have you been able to nail down the service(s) that is/are affected by this issue - is it `SERVICES=dynamodb`, by any chance? (We believe this should have been fixed recently, but perhaps there's some other issue remaining..)

**Comment by ktamas77** [comment]
> Hi @ktamas77 , thanks for the update. Can you please share the tag/age of the image you're using (I'm assuming you pulled the `latest` image, not the tagged version)?

Of course (you right, I pulled the latest):

```
localstack/localstack                                               latest            3ef1ed8c20d2   8 hours ago    652MB
```

> Have you been able to nail down the service(s) that is/are affected by this issue - is it `SERVICES=dynamodb`, by any chance? (We believe this should have been fixed recently, but perhaps there's some other issue remaining..)

I can try a process of elimination to see if there is a particular service which enters to the infinite loop at restart when persistent volume is enabled.

**Comment by pchico83** [comment]
I am seeing the same issue just with sqs and sns services on 0.12.4.
Using `DATA_DIR`, first boot works ok, then I restart the localstack container and it enters the `Waiting for all LocalStack services to be ready` loop
